pennylane/gradients/pulse_gradient.py

```diff
@@ -608,12 +674,24 @@ def _expval_stoch_pulse_grad(tape, argnum, num_split_times, key, shots, use_broa
             continue
 
         key, _key = jax.random.split(key)
-        cjacs, _tapes, avg_prefactor, psr_coeffs = _generate_tapes_and_cjacs(
-            tape, idx, _key, num_split_times, use_broadcasting
-        )
+        op_id = tape.get_operation(idx)
+        op, *_ = op_id
+        if not isinstance(op, ParametrizedEvolution):
+            raise ValueError(
+                "stoch_pulse_grad does not support differentiating parameters of "
+                "other operations than pulses."
+            )
+        if isinstance(op.H, HardwareHamiltonian):
+            # Treat HardwareHamiltonians separately because they have a reordering function
+            _tapes, data = _tapes_data_hardware(tape, op_id, key, num_split_times, use_broadcasting)
```

Comment:
how come the hardware hamiltonian case is that much simpler? (i.e. not returning `int_prefactor, psr_coeffs`)

Reply:
It looks simpler, but only because `_generate_tapes_and_cjacs` and `_tapes_data_hardware` are actually _not_ on the same function level. `_tapes_data_hardware` uses `_generate_tapes_and_cjacs` itself, so there is an entire additional nesting level for `HardwareHamiltonian`